### PR TITLE
DOC: transition of examples from 2 to 3. Closes #7366

### LINKED ISSUE #7366: Transition code examples to Python 3
The Scipy examples and tutorials currently contain a number of examples that won't run on Python 3.
It could make sense to just move all of them to Python 3 syntax, and enforce this by running the REFGUIDE_CHECK on travis on Python3 instead of Python2.

### PR DESCRIPTION
**Pull request by andyfaff**
This is going to be a work in progress. There seems to be a lot of changes required.

### CODE REVIEW

**Comment by andyfaff** [comment]
I'm working by fixing the examples on my computer, not travis. Not sure how to deal with the docstring in cython code. The following fails because there isn't any data:

```
File "build/testenv/lib/python3.6/site-packages/scipy/spatial/ckdtree.cpython-36m-darwin.so", line ?, in cKDTree.query
Failed example:
    tree = cKDTree(data)
Exception raised:
    Traceback (most recent call last):
      File "/Users/anz/miniconda3/envs/dev3/lib/python3.6/doctest.py", line 1330, in __run
        compileflags, 1), test.globs)
      File "<doctest cKDTree.query[0]>", line 1, in <module>
        tree = cKDTree(data)
    NameError: name 'cKDTree' is not defined
```

**Comment by andyfaff** [comment]
Can someone check 989779e (csgraph.rst)? Not sure if I'm handling the strings correctly.

**Comment by andyfaff** [comment]
@pv who should I ask about the cKTtree example? There is a docstring test in cython code, but it would never actually work because there's no data to use with it.
It's in `scipy/spatial/ckdtree.pyx` starting on line 703.

**Comment by andyfaff** [comment]
Spoke too soon. There is a csgraph entry failing, there must be a word in the word list that has an accent.

**Comment by rgommers** [comment]
> @pv who should I ask about the cKDtree example? 

Yeah that needs fixing. Example was written by @rainwoodman in commit 501e009a3cc. Seems fine to skip the example in refguide-check for now and open a new issue for it.

**Comment by andyfaff** [comment]
I might try and create an example based on https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.KDTree.query.html

**Comment by andyfaff** [comment]
Ok, one step further. I lifted the example for cKDTtree and it works. However, refguide check is still failing with:

```
==========
scipy.misc
==========

Non-deprecated objects in __all__: 5
Objects in refguide: 23

Deprecated objects in refguide::

    comb
    factorial
    factorial2
    factorialk
    info
    logsumexp
    pade
    source
    who
ERROR: objects in refguide but not in scipy.misc.__all__::

    bytescale
    fromimage
    imfilter
    imread
    imresize
    imrotate
    imsave
    imshow
    toimage

This issue should likely be fixed by removing these objects
from the function listing in __init__.py for this module
or adding them to __all__.

scipy.misc.bytescale
--------------------

scipy.misc.bytescale has no docstring

scipy.misc.toimage
------------------

scipy.misc.toimage has no docstring

scipy.misc.imfilter
-------------------

scipy.misc.imfilter has no docstring

scipy.misc.imresize
-------------------

scipy.misc.imresize has no docstring

scipy.misc.imsave
-----------------

scipy.misc.imsave has no docstring

scipy.misc.imshow
-----------------

scipy.misc.imshow has no docstring

scipy.misc.imrotate
-------------------

scipy.misc.imrotate has no docstring

scipy.misc.imread
-----------------

scipy.misc.imread has no docstring

scipy.misc.fromimage
--------------------

scipy.misc.fromimage has no docstring

```

I'll have to defer to someone else here, don't know what to do here.

**Comment by andyfaff** [comment]
Figured it out for myself - I had to install Pillow. The refguide tests now pass on my computer, lets see what Travis thinks.

**Comment by andyfaff** [comment]
The reviewer should concentrate on changes made to csgraph and ckdtree. The former may not be the best way to deal with unicode.

File: scipy/spatial/ckdtree.pyx
Comment (rainwoodman) [inline scipy/spatial/ckdtree.pyx]: Is there a better number construct for doing this? I see you used numpy.c_ later.

**Comment by andyfaff** [comment]
Rebased on master.

**Comment by andyfaff** [comment]
I squashed all the commits into 1. There is a lot of work going on in .travis.yml and it was playing havoc with rebasing on master all the time.

**Comment by andyfaff** [comment]
Test failure seems unrelated.

**Review by larsoner — APPROVED** [review]
Changes LGTM

File: doc/source/tutorial/csgraph.rst
Comment (perimosocordiae) [inline doc/source/tutorial/csgraph.rst]: For simplicity, I would prefer to just decode these unicode strings to ascii (`bytes`) and leave the rest essentially unchanged.

File: tools/win32/detect_cpu_extensions_wine.py
Comment (perimosocordiae) [inline tools/win32/detect_cpu_extensions_wine.py]: Looks like we're losing a linebreak here, after the separator. Maybe do:
```python
print("", "-" * 80,
      "Checked all binary files for CPU extension codes. "
      "Found the following codes:", sep="\n")
```

Comment (ev-br) [inline tools/win32/detect_cpu_extensions_wine.py]: I'm surprised we're checking these, maybe it's easier to just skip checking the docstrings of anything under tools/ ? 

Also this code in particular might be dead code --- we've dropped building under wine around scipy 0.17.0 IIRC.

Comment (andyfaff) [inline tools/win32/detect_cpu_extensions_wine.py]: I'm not sure if this is ever used. It would only ever work on Python 2 because of the print statement. The same argument applies to `tools/osx/build.py`
The way I started off this PR was to check for "print " in my IDE to see where the remaining print problems were in docstrings. I also ended up changing things like these.

Comment (andyfaff) [inline tools/win32/detect_cpu_extensions_wine.py]: I'll make the change anyway. If the file's not used it can be removed in a future commit.

Comment (ev-br) [inline tools/win32/detect_cpu_extensions_wine.py]: Last time I checked, the release scripts had some py2-isms, so that the release process only worked on python 2.x.

Anyway, I suggest focussing this PR on the docstrings which are actually user-facing, and looking into removing dead code from `tools/` in a separate PR (which will need to be vetted by Ralf or Pauli).

BTW, the bulk of the changes here LGTM.

Comment (ev-br) [inline tools/win32/detect_cpu_extensions_wine.py]: Sorry, cross-talking. I'm fine with keeping this change, too.

File: doc/source/tutorial/csgraph.rst
Comment (ev-br) [inline doc/source/tutorial/csgraph.rst]: Not sure about this --- the comment just switches off the check; is there some reason why the number becomes wrong under python 3?

**Review by ev-br — COMMENTED** [review]
LGTM modulo one query.

File: doc/source/tutorial/csgraph.rst
Comment (andyfaff) [inline doc/source/tutorial/csgraph.rst]: It depends on how long the word list on any given computer. I have 1135 three letter words on my Mac.

File: doc/source/tutorial/csgraph.rst
Comment (andyfaff) [inline doc/source/tutorial/csgraph.rst]: There are unicode characters in the dictionary (probably accents), so would need to decode to UTF-8.
I thought about using that approach, but when the words are printed out they are changed to bytes, i.e. `and` becomes `b'and'`.

File: doc/source/tutorial/csgraph.rst
Comment (andyfaff) [inline doc/source/tutorial/csgraph.rst]: This then changes the rest of the answers. So it makes the refguide check dependent on a given POSIX dictionary which is fragile.

File: doc/source/tutorial/csgraph.rst
Comment (andyfaff) [inline doc/source/tutorial/csgraph.rst]: The main issue here is that there are unicode characters in the travis word list. e.g.

```
>>> 'aèr'.encode('ascii')
UnicodeEncodeError: 'ascii' codec can't encode character '\xe8' in position 1: ordinal not in range(128)
>>> len('aèr'.encode('utf-8'))
4
>>> len('aer'.encode('utf-8'))
3
```

So it's not possible to encode to `ascii` unless one ignores those characters. It's not a good idea to encode to `UTF-8` because different words require different numbers of bytes:
```
>>> word_list = [word.encode('UTF-8') for word in ['aèr', 'aer', 'èèè']]
>>> word_list = np.asarray(word_list)
>>> word_list.dtype
dtype('S6')
>>> print(word_list)
[b'a\xc3\xa8r' b'aer' b'\xc3\xa8\xc3\xa8\xc3\xa8']
>>> word_bytes = np.ndarray((len(word_list), word_list.itemsize),
...                                                dtype='uint8',
...                                                buffer=word_list.data)
>>> print(word_bytes)
array([[ 97, 195, 168, 114,   0,   0],
       [ 97, 101, 114,   0,   0,   0],
       [195, 168, 195, 168, 195, 168]], dtype=uint8)
```
The words that have no 'fancy characters' only take three bytes and the rest of the row is padded zero. The total number of columns in `word_bytes` depends on the word that has the largest number of bytes in its UTF-8 representation.

 The way that the PR works at the moment is as follows:

- Each of the 3 letter words in `word_list` is UTF-8, that's the default encoding in my locale.
- `word_list = np.asarray(word_list)` puts them into a numpy array which holds 3 unicode characters. `word_list.itemsize` is 12, a UTF-8 character has a maximum size of 4 bytes.
-   In the conversion from `word_list` to `word_bytes` we have to reserve 12 bytes for each word (number of columns).
```
 >>> word_bytes = np.ndarray((word_list.size, word_list.itemsize),
    ...                         dtype='uint8',
    ...                         buffer=word_list.data)
```
- I'm assuming that all characters can be represented by single bytes, so only the first byte of every character is required, so to represent each letter take every 4th byte: `word_bytes = word_bytes[:, ::4]`, which is equal to `word_bytes[:, ::word_list.itemsize // 3`. This approach allows you to create a byte array for each of `['aèr', 'aer', 'èèè']`

**Pull request merged**
